polygeist/lib/polygeist/Passes/ConvertToLLVMABI.cpp

```diff
@@ -0,0 +1,123 @@
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This pass currently only handles SYCL Kernel functions.
+// This pass converts all arguments with MemRef type to LLVM pointer type,
+// and replace all uses of the original argument with a
+// `polygeist.pointer2memref` of the new argument.
+//
+//===----------------------------------------------------------------------===//
+
+#include "PassDetails.h"
+#include "mlir/Conversion/LLVMCommon/Pattern.h"
+#include "mlir/Conversion/LLVMCommon/TypeConverter.h"
+#include "mlir/Dialect/GPU/IR/GPUDialect.h"
+#include "mlir/Dialect/LLVMIR/LLVMDialect.h"
+#include "mlir/Dialect/MemRef/IR/MemRef.h"
+#include "mlir/Transforms/DialectConversion.h"
+#include "llvm/Support/Debug.h"
+
+#define DEBUG_TYPE "convert-to-llvm-abi"
+
+using namespace mlir;
+
+namespace {
+
+class GPUFuncLowering : public OpRewritePattern<gpu::GPUFuncOp> {
+public:
+  using OpRewritePattern<gpu::GPUFuncOp>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(gpu::GPUFuncOp op,
+                                PatternRewriter &rewriter) const override {
+    assert(op->getAttr("llvm.cconv") ==
+               mlir::LLVM::CConvAttr::get(op.getContext(),
+                                          LLVM::cconv::CConv::SPIR_KERNEL) &&
+           "Expecting SYCL Kernel");
+    LLVM_DEBUG(llvm::dbgs() << "ConvertToLLVMABIPass: GPUFuncLowering: "
+                            << op.getName() << "\n");
+
+    // Notify MLIR we're updating the function in place
+    rewriter.startRootUpdate(op);
+
+    FunctionType funcTy = op.getFunctionType();
+    TypeConverter::SignatureConversion conversion(funcTy.getNumInputs());
+
+    rewriter.setInsertionPointToStart(&op.getBody().front());
+    Region &funcBody = op.getBody();
+    SmallVector<BlockArgument, 8> Args(funcBody.getArguments().begin(),
+                                       funcBody.getArguments().end());
+    for (auto &en : llvm::enumerate(Args)) {
+      Value arg = en.value();
+      auto MT = arg.getType().dyn_cast<MemRefType>();
+      if (!MT) {
+        conversion.addInputs(en.index(), {arg.getType()});
+        continue;
+      }
+
+      LLVM_DEBUG(llvm::dbgs() << "  Replace argument " << en.index() << ": \""
+                              << arg << "\"");
+      Type PT = LLVM::LLVMPointerType::get(MT.getElementType(),
+                                           MT.getMemorySpaceAsInt());
+      Value newArg = funcBody.insertArgument(en.index(), PT, arg.getLoc());
+      LLVM_DEBUG(llvm::dbgs() << " -> \"" << newArg << "\"\n");
+      conversion.addInputs(en.index(), {PT});
+
+      auto Ptr2Memref = rewriter.create<polygeist::Pointer2MemrefOp>(
+          arg.getLoc(), MT, newArg);
+
+      arg.replaceAllUsesWith(Ptr2Memref);
+      funcBody.eraseArgument(en.index() + 1);
+    }
```

Comment:
Missing: structures should be passed as pointers to a struct with the `byval` attribute.

Reply:
@etiotto is working on a different PR for this 😃